embassy-net/src/dns.rs

```diff
@@ -0,0 +1,114 @@
+//! DNS socket with async support.
+use core::cell::RefCell;
+use core::future::poll_fn;
+use core::mem;
+use core::task::Poll;
+
+use embassy_net_driver::Driver;
+use heapless::Vec;
+use managed::ManagedSlice;
+use smoltcp::iface::{Interface, SocketHandle};
+pub use smoltcp::socket::dns::DnsQuery;
+use smoltcp::socket::dns::{self, GetQueryResultError, StartQueryError, MAX_ADDRESS_COUNT};
+pub use smoltcp::wire::{DnsQueryType, IpAddress};
+
+use crate::{SocketStack, Stack};
+
+/// Errors returned by DnsSocket.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// No available query slot
+    NoFreeSlot,
```

Comment:
`NoFreeSlot` could async-wait until there is one, instead of failing.

Reply:
Since the query slots are provided in new() and DnsSocket::query is &mut self, would it actually be possible to have a waiter unless query() changes to &self?

Reply:
ah good point, seems like no.

I was thinking about the case of creating one DnsSocket and then sharing it "system-wide" to do DNS queries for all tasks. You can't create two DnsSockets, the first one will process and discard all packets for the 2nd one.

Perhaps the socket should be owned by the Stack, and the stack should have a `&self`  method for doing a DNS query? so multiple tasks can do multiple queries concurrently?

Reply:
> ah good point, seems like no.
> 
> I was thinking about the case of creating one DnsSocket and then sharing it "system-wide" to do DNS queries for all tasks. You can't create two DnsSockets, the first one will process and discard all packets for the 2nd one.
> 

Is this a limit internally in smoltcp? 

> Perhaps the socket should be owned by the Stack, and the stack should have a `&self` method for doing a DNS query? so multiple tasks can do multiple queries concurrently?

Given the above, yes maybe that's a better approach, I can give it a go.

Reply:
> Is this a limit internally in smoltcp?

Yes, the packets are given to the first socket that accepts it https://github.com/smoltcp-rs/smoltcp/blob/739e7a58a9a41ce027eaf1e8857dc507b70f4756/src/iface/interface/mod.rs#L1380
and the DnsSocket accepts all DNS packets https://github.com/smoltcp-rs/smoltcp/blob/739e7a58a9a41ce027eaf1e8857dc507b70f4756/src/socket/dns.rs#L340

Reply:
I've changed it to use a single DnsSocket (separate type just for convenience in imports and 'hiding' the stack impl), implemented on the Stack type, which I think makes more sense after implementing it, because it ties into the DNS server updates nicely.

I didn't add the waker in the case of new slots, and created slots equal to the number of sockets (SOCK on the resources). We could add wakers as well, but I'm wondering if it's really needed, as it would just 'postpone' the problem of no free slots by N wakers, which I think just complicates types and config. Or am I missing something?

Reply:
Just 1 waker should be enough. If there's two tasks waiting they'll fight over the WakerRegistration, just like in `embassy-sync` stuff, but it should all still work. This'd work for an infinite amount of tasks waiting.